### DISCUSSION TITLE
LIBCSDK - PUT request with Integer Array for a device service

### DISCUSSION THREAD

**Discussion by vijay-hcl-epl** [edgexfoundry/community#39] [Q&A] [answered]
Hi EdgeX team,
I am developing a C device service and I have a requirement to send multiple parameters in the PUT request to that device service. To achieve this, I used **"valueType": "Int8Array"** in the device profile. Then I tried sending curl command as shown below.

`curl -X 'PUT' -H "Content-Type: application/json" 'http://localhost:59882/api/v2/device/name/dev1/wrres' -d '{"wrres" : "[435]"}' | json_pp` . It gives the response **failed to send a http request**.  Also the device service got crashed with segmentation fault. It is crashing even before it reaches the PUT handler in device service. When I enabled backtrace I could see something like below 

> /lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7f595f970520]
> /lib/x86_64-linux-gnu/libc.so.6(+0x1afcb8)[0x7f595faddcb8]
> /lib/x86_64-linux-gnu/libc.so.6(+0x8fa68)[0x7f595f9bda68]
> /lib/x86_64-linux-gnu/libc.so.6(__isoc99_sscanf+0xd3)[0x7f595f9903a3]
> /lib/libcsdk.so(edgex_device_handler_device_namev2+0xa09)[0x7f595fb742f9]
> /lib/libcsdk.so(+0x383eb)[0x7f595fb8e3eb]
> /lib/x86_64-linux-gnu/libmicrohttpd.so.12(+0xaa7c)[0x7f595f827a7c]
> /lib/x86_64-linux-gnu/libmicrohttpd.so.12(+0xcb48)[0x7f595f829b48]
> /lib/x86_64-linux-gnu/libmicrohttpd.so.12(+0xf440)[0x7f595f82c440]
> 

To double check, I executed the random device service given as a example in https://github.com/edgexfoundry/device-sdk-c/tree/508af3997b99b63bbc6444d3a4c84739e7448765/src/c/examples/random. I modified the device profile with **"valueType": "Int8Array"** . It also gave the same behavior(device service crashed) when I sent PUT request with integer array.

Please let me know how to send PUT request with integer array or any other way to send multiple parameters in the PUT request.

Thank you,
Vijay

**Comment by iain-anderson** [comment, marked as answer]
That's a C-SDK bug. It is still expecting the old format for array data where each element is a string. You can work around this by sending` '{"wrres" : "[\"435\"]"}'`
I'm working on an update that changes some of the parsing code - I'll test that this works properly as part of that update (including being more robust with unexpected data)

_Edit: actually 435 should fail as an Int8 value_

**Reply by vijay-hcl-epl** [reply]
Thank you Iain for the reply..!! It worked with the way you suggested. Also, when we use array of data, should we use the APIs iot_data_array_iter, iot_data_array_iter_next, iot_data_array_iter_value to parse and extract the data in PUT handler? Please let us know

Thanks,
Vijay.

**Reply by iain-anderson** [reply]
Yes, the pattern is
```
iot_data_array_iter_t iter;
iot_data_array_iter (arr, &iter);
while (iot_data_array_iter_next (&iter))
{
  uint32_t val = *((uint32_t *) iot_data_array_iter_value (&iter));
  // etc
}
```
with whatever type is appropriate substituted for uint32_t
Alternatively you can use `iot_data_address()` to directly access the base address of the contained C array. `iot_data_array_size()` gives its size in bytes, `iot_data_array_length()` gives the number of elements

**Reply by vijay-hcl-epl** [reply]
Thanks a lot Iain..!!
